src/axom/quest/detail/MarchingCubesImpl.hpp

```diff
@@ -0,0 +1,756 @@
+// Copyright (c) 2017-2023, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#include "axom/core/execution/execution_space.hpp"
+#include "axom/quest/MarchingCubes.hpp"
+#include "axom/quest/detail/marching_cubes_lookup.hpp"
+#include "axom/primal/geometry/Point.hpp"
+#include "axom/primal/constants.hpp"
+#include "axom/mint/execution/internal/structured_exec.hpp"
+#include "conduit_blueprint.hpp"
+#include "axom/fmt.hpp"
+
+namespace axom
+{
+namespace quest
+{
+namespace detail
+{
+namespace marching_cubes
+{
+template <typename T, int DIM, typename U>
+static void add_to_StackArray(axom::StackArray<T, DIM>& a, U b)
+{
+  for(int d = 0; d < DIM; ++d) a[d] += b;
+}
+
+//!@brief Reverse the order of a StackArray.
+template <typename T, int DIM>
+static void reverse(axom::StackArray<T, DIM>& a)
+{
+  for(int d = 0; d < DIM / 2; ++d)
+  {
+    std::swap(a[d], a[DIM - 1 - d]);
+  }
+  return;
+}
+
+/*!
+  @brief Computations for MarchingCubesSingleDomain
+
+  Spatial dimension templating is here, to keep out of higher level
+  classes MarchCubes and MarchingCubesSingleDomain.
+*/
+template <int DIM, typename ExecSpace>
+class MarchingCubesImpl : public MarchingCubesSingleDomain::ImplBase
+{
+public:
+  using Point = axom::primal::Point<double, DIM>;
+  using MIdx = axom::StackArray<axom::IndexType, DIM>;
+  using LoopPolicy = typename execution_space<ExecSpace>::loop_policy;
+  using ReducePolicy = typename execution_space<ExecSpace>::reduce_policy;
+  static constexpr auto MemorySpace = execution_space<ExecSpace>::memory_space;
+  /*!
+    @brief Initialize data to a blueprint domain.
+    @param dom Blueprint structured mesh domain
+    @param coordsetPath Where coordinates are in dom
+    @param fcnPath Where nodal function is in dom
+    @param maskPath Where cell mask function is in dom
+
+    Set up views to domain data and allocate other data to work on the
+    given domain.
+  */
+  AXOM_HOST void initialize(const conduit::Node& dom,
+                            const std::string& coordsetPath,
+                            const std::string& fcnPath,
+                            const std::string& maskPath) override
+  {
+    clear();
+
+    // Data sizes
+    const conduit::Node& dimsNode =
+      dom.fetch_existing("topologies/mesh/elements/dims");
+    for(int d = 0; d < DIM; ++d)
+    {
+      m_bShape[d] = dimsNode[d].as_int();
+    }
+    m_cShape = m_bShape;
+    reverse(m_cShape);
+    // This should work but breaks gcc11 on 64-bit linux:
+    // m_pShape = m_cShape + 1;
```

Comment:
How is not adding 1 still work? is that what the `add_to_StackArray(m_pShape, 1)` is doing?

Reply:
It seems to insist that the `operator+` returns a pointer and complains that the pointer can't be cast into a `StackArray`.  The `add_to_StackArray` does the same thing, but returns `void`.  Weird.